pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/periodictask/ControllerPeriodicTask.java

```diff
@@ -0,0 +1,101 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.controller.helix.core.periodictask;
+
+import com.linkedin.pinot.controller.helix.core.PinotHelixResourceManager;
+import com.linkedin.pinot.core.periodictask.BasePeriodicTask;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The base periodic task for pinot controller only. It uses <code>PinotHelixResourceManager</code> to determine
+ * which table resources should be managed by this Pinot controller.
+ */
+public abstract class ControllerPeriodicTask extends BasePeriodicTask {
+  public static final Logger LOGGER = LoggerFactory.getLogger(ControllerPeriodicTask.class);
+  protected final PinotHelixResourceManager _pinotHelixResourceManager;
+  private boolean _amILeader;
+  private static final int DEFAULT_INITIAL_DELAY_IN_SECOND = 120;
+
+  public ControllerPeriodicTask(String taskName, long runFrequencyInSeconds,
+      PinotHelixResourceManager pinotHelixResourceManager) {
+    this(taskName, runFrequencyInSeconds, DEFAULT_INITIAL_DELAY_IN_SECOND, pinotHelixResourceManager);
+  }
+
+  public ControllerPeriodicTask(String taskName, long runFrequencyInSeconds, long initialDelayInSeconds,
+      PinotHelixResourceManager pinotHelixResourceManager) {
+    super(taskName, runFrequencyInSeconds, initialDelayInSeconds);
+    _pinotHelixResourceManager = pinotHelixResourceManager;
+    _amILeader = false;
+  }
+
+  @Override
+  public void init() {
+  }
+
+  @Override
+  public void run() {
+    if (!_pinotHelixResourceManager.isLeader()) {
+      skipLeaderTask();
+    } else {
+      List<String> allTableNames = _pinotHelixResourceManager.getAllTables();
+      processLeaderTask(allTableNames);
+    }
+  }
+
+  private void skipLeaderTask() {
+    if (_amILeader) {
+      onBecomeNotLeader();
+    }
+    _amILeader = false;
+    printSkipMessage();
+  }
+
+  private void processLeaderTask(List<String> allTableNames) {
+    if (!_amILeader) {
+      onBecomeLeader();
+    }
+    _amILeader = true;
+    process(allTableNames);
+  }
+
+  /**
+   * Does the following logic when losing the leadership. This should be done only once during leadership transition.
+   */
+  public void onBecomeNotLeader() {
+    LOGGER.info("Lost leadership");
+  }
+
+  /**
+   * Does the following logic when becoming lead controller. This should be done only once during leadership transition.
+   */
+  private void onBecomeLeader() {
+    LOGGER.info("Became leader");
+  }
+
+  /**
+   * Processes the periodic task as lead controller.
+   * @param allTableNames List of all the table names
+   */
+  public abstract void process(List<String> allTableNames);
+
+  /**
+   * Prints message of skipping the periodic task.
+   */
+  public abstract void printSkipMessage();
```

Comment:
You should not need this method to be abstract. The base class can log this message, right?